src/protocol/modulus_conversion/convert_shares.rs

```diff
@@ -134,28 +150,47 @@ where
 /// Propagates errors from convert shares
 /// # Panics
 /// Propagates panics from convert shares
-pub async fn convert_all_bits<F, C, S>(
+pub async fn convert_all_bits<'a, F, C, S, I>(
     ctx: &C,
-    locally_converted_bits: &[Vec<BitConversionTriple<S>>],
-) -> Result<Vec<Vec<S>>, Error>
+    mut locally_converted_bits: I,
+    num_bits: u32,
+    num_multi_bits: u32,
+) -> Result<impl Iterator<Item = Vec<Vec<S>>> + 'a, Error>
 where
     F: Field,
     C: Context<F, Share = S>,
-    S: ArithmeticSecretSharing<F>,
+    S: ArithmeticSecretSharing<F> + 'a,
+    I: Iterator<Item = Vec<BitConversionTriple<S>>>,
 {
-    let futures = locally_converted_bits
-        .iter()
+    let num_records = locally_converted_bits.size_hint().0;
+
+    let count = ((num_bits + num_multi_bits - 1) / num_multi_bits) as usize;
+    let mut futures = Vec::with_capacity(count);
+    futures.resize_with(count, || Vec::with_capacity(num_records));
+
+    locally_converted_bits
+        .by_ref()
         .enumerate()
-        .map(|(bit_num, one_column)| {
-            convert_bit_list(
-                ctx.narrow(&ModulusConversion(bit_num.try_into().unwrap()))
-                    .set_total_records(one_column.len()),
-                one_column,
-            )
-        })
-        .collect::<Vec<_>>();
-    let converted_shares = try_join_all(futures).await?;
-    Ok(converted_shares)
+        .for_each(|(idx, record)| {
+            (0..count).for_each(|slice_num| {
+                let first_bit =
+                    num_multi_bits * <usize as TryInto<u32>>::try_into(slice_num).unwrap();
+                let last_bit = std::cmp::min(first_bit + num_multi_bits, num_bits) as usize;
+                futures[slice_num].push(convert_bit_list(
+                    ctx.narrow(&ModulusConversion(first_bit))
+                        .set_total_records(num_records),
+                    record[first_bit as usize..last_bit].to_vec(),
+                    RecordId::from(idx),
+                ));
```

Comment:
I don't think we want to call "convert_bit_list" on ALL of the bits at once. When it comes time to upgrade to malicious security, we will want to just convert the bits that are inputs to the next stage of the sort algorithm, use them, call validate, and then we can call "convert_bit_list" on the next batch, as they are inputs to the next stage, and need to be validated along with the circuit where they are used.